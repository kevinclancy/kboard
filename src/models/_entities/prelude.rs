//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.13

pub use super::boards::Entity as Boards;
pub use super::replies::Entity as Replies;
pub use super::reply_statuses::Entity as ReplyStatuses;
pub use super::threads::Entity as Threads;
pub use super::users::Entity as Users;
