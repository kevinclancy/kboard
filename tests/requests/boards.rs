use insta::{assert_debug_snapshot};
use kboard::{
    app::App,
    models::users,
    controllers::boards::RepliesResponse as RepliesResponse,
    controllers::boards::ThreadsResponse as ThreadsResponse
};
use loco_rs::testing::prelude::*;
use serial_test::serial;

// TODO: see how to dedup / extract this to app-local test utils
// not to framework, because that would require a runtime dep on insta
macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("board_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_get_boards() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();
        let response = request.get("/api/boards").await;
        assert_eq!(
            response.status_code(),
            200,
            "Boards request should succeed"
        );

        assert_debug_snapshot!(response.text());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_threads() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();
        let response = request.get("/api/boards/1/threads").await;
        assert_eq!(
            response.status_code(),
            200,
            "Threads request should succeed"
        );

        assert_debug_snapshot!(response.text());
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_post_thread() {
    configure_insta!();

    let payload = serde_json::json!({
        "title": "I was startled by a Pika.",
        "initial_reply_text": "The Pika's cry caught me off guard.",
    });

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();

        // Get user1 from fixtures and generate JWT token
        let user1 = users::Model::find_by_pid(&ctx.db, "11111111-1111-1111-1111-111111111111").await.unwrap();
        let jwt_secret = ctx.config.get_jwt_config().unwrap();
        let token = user1
            .generate_jwt(&jwt_secret.secret, jwt_secret.expiration)
            .unwrap();

        let response = request
            .post("/api/boards/1/threads")
            .add_header("authorization", format!("Bearer {}", token))
            .json(&payload)
            .await;

        assert_eq!(response.status_code(), 200);

        let response_text = response.text();
        assert_debug_snapshot!(response_text);

        let response = request.get("/api/boards/1/threads/5/replies?page_size=1&page_number=0").await;
        assert_eq!(response.status_code(), 200);

        let replies_response: RepliesResponse = response.json();
        assert_eq!(replies_response.replies.len(), 1);
        assert_eq!(replies_response.replies[0].body, "The Pika's cry caught me off guard.");
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_thread() {
    configure_insta!();

    request::<App, _, _>(|request, ctx| async move {
        seed::<App>(&ctx).await.unwrap();

        // Get user1 from fixtures and generate JWT token
        let user1 = users::Model::find_by_pid(&ctx.db, "11111111-1111-1111-1111-111111111111").await.unwrap();
        let jwt_secret = ctx.config.get_jwt_config().unwrap();
        let token = user1
            .generate_jwt(&jwt_secret.secret, jwt_secret.expiration)
            .unwrap();

        let del_response = request
            .delete("/api/boards/1/threads/1")
            .add_header("authorization", format!("Bearer {}", token))
            .await;

        assert_eq!(del_response.status_code(), 200);


        let get_response = request.get("/api/boards/1/threads").await;
        assert_eq!(get_response.status_code(), 200);
        assert_debug_snapshot!(get_response.text());

        let threads_response: ThreadsResponse = get_response.json();
        assert!(threads_response.threads.iter().find(|t| (t.id == 1)).is_none());
    })
    .await;
}