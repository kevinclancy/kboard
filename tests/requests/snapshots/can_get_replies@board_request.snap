---
source: tests/requests/boards.rs
expression: response.text()
---
"{\"replies\":[{\"id\":1,\"body\":\"I was runnning. A seagull startled me, and then I fell on my face.\",\"reply_to\":null,\"thread_id\":1,\"poster\":1,\"poster_username\":\"user1\",\"poster_is_banned\":false,\"updated_at\":\"2023-11-12T12:34:56.789Z\",\"reply_status\":1},{\"id\":5,\"body\":\"Oh wow. That's terrible!\",\"reply_to\":[1,\"I was runnning. A seagull startled me, and then I fell on my face.\",1],\"thread_id\":1,\"poster\":2,\"poster_username\":\"user2\",\"poster_is_banned\":false,\"updated_at\":\"2023-11-12T12:34:56.790Z\",\"reply_status\":1}],\"total_count\":1,\"thread_title\":\"Startled by a seagull.\",\"board_name\":\"Broken Jaw Stories\"}"
